### PR TITLE
Sign up

### PR DESCRIPTION
**Pull request by jxczz2**
## Summary

Sign up is added with their respective validations.

## Known Issues

No problem.

## Trello Card

https://trello.com/c/fJ5jAbKr

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -29,6 +30,8 @@ const init = () => {
 
   app.use('/docs', express.static(path.join(__dirname, 'docs')));
 
+  app.use(expressValidator());
```
Comment (picollomartin) [inline app.js]: This is not needed

File: app/controllers/user.controller.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  error = require('../errors'),
+  logger = require('../logger'),
+  User = require('../models').User;
+
+exports.userCreate = (req, res, next) => {
+  const user = req.body;
+
+  user.password = bcrypt.hashSync(user.password, salt);
+
+  User.create(user)
```
Comment (picollomartin) [inline app/controllers/user.controller.js]: This should be done in model or in a service

File: app/controllers/user.controller.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      logger.info(`User ${createdUser.dataValues.name} created correctly.`);
+      res
+        .status(201)
+        .send('User created Correctly')
+        .end();
+    })
+    .catch(err => {
+      logger.info(`${user.name} user no created.`);
+      logger.error(err);
+      if (err.name === 'SequelizeUniqueConstraintError') {
```
Comment (picollomartin) [inline app/controllers/user.controller.js]: This validation is for model or service, not in controller.

File: app/controllers/user.controller.js (hunk 2 above)
Comment (picollomartin) [reply app/controllers/user.controller.js]: Return promise too

File: app/controllers/user.controller.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+'use strict';
+
+const bcrypt = require('bcryptjs'),
+  saltRounds = 10,
```
Comment (picollomartin) [inline app/controllers/user.controller.js]: Only as a comment, we use a constant file

File: app/controllers/user.controller.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  const user = req.body;
+
+  user.password = bcrypt.hashSync(user.password, salt);
+
+  User.create(user)
+    .then(createdUser => {
+      logger.info(`User ${createdUser.dataValues.name} created correctly.`);
+      res
+        .status(201)
+        .send('User created Correctly')
```
Comment (picollomartin) [inline app/controllers/user.controller.js]: Empty response is ok, remove message

File: app/middlewares/signup-validations.middleware.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+exports.signUpValidations = [
+  check('name')
+    .not()
+    .isEmpty()
+    .withMessage('Name is required'),
+
+  check('lastName')
+    .not()
+    .isEmpty()
+    .withMessage('Lastname is required'),
```
Comment (picollomartin) [inline app/middlewares/signup-validations.middleware.js]: Some duplicated logic beetwen validations

File: app/middlewares/signup-validations.middleware.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    .isLength({ min: 8 })
+    .withMessage('minimum 8 characters'),
+
+  (req, res, next) => {
+    const errors = validationResult(req);
+    const msg = !errors.isEmpty()
+      ? errors.array()
+      : !/@wolox+(\.\w{2,3})+$/.test(req.body.email) ? 'email domain incorrect' : 'OK';
+
+    if (msg === 'OK') {
```
Comment (picollomartin) [inline app/middlewares/signup-validations.middleware.js]: Do something generic for check validations, this hardcoded message is tricky too

File: app/controllers/user.controller.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
+'use strict';
```
Comment (picollomartin) [inline app/controllers/user.controller.js]: check naming of file, use `users.js`

File: app/middlewares/signup-validations.middleware.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
+const { check, validationResult } = require('express-validator/check'),
```
Comment (picollomartin) [inline app/middlewares/signup-validations.middleware.js]: We use camelCase for name files, .middleware is not needed

File: app/controllers/user.controller.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  saltRounds = 10,
+  salt = bcrypt.genSaltSync(saltRounds),
+  error = require('../errors'),
+  logger = require('../logger'),
+  User = require('../models').User;
+
+exports.userCreate = (req, res, next) => {
+  const user = req.body;
+
+  user.password = bcrypt.hashSync(user.password, salt);
```
Comment (picollomartin) [inline app/controllers/user.controller.js]: What happen if body is empty ?

File: app/middlewares/signup-validations.middleware.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  (req, res, next) => {
+    const errors = validationResult(req);
+    const msg = !errors.isEmpty()
+      ? errors.array()
+      : !/@wolox+(\.\w{2,3})+$/.test(req.body.email) ? 'email domain incorrect' : 'OK';
+
+    if (msg === 'OK') {
+      next();
+    } else {
+      next(error.defaultError(msg));
```
Comment (picollomartin) [inline app/middlewares/signup-validations.middleware.js]: Default error ? 4xx for badRequest, 5xx for internal server error and 2xx ok

**Review by picollomartin — CHANGES_REQUESTED** [review]
Check responsibilities, controller should call services and/or models for do things (create user for example, call an external API, etc)

File: app/controllers/users.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+const error = require('../errors'),
+  User = require('../models').User;
+
+exports.userCreate = (req, res, next) => {
+  User.createUser(req.body)
```
Comment (picollomartin) [inline app/controllers/users.js]: Return promise here, what happen if i send additional properties in body ?

File: app/models/user.js [hunk 13]
```diff
@@ -1,19 +1,41 @@
...
     },
     {
-      underscore: true,
+      underscored: true,
       freezeTableName: true
     }
   );
 
+  User.createUser = user => {
+    return new Promise((resolve, reject) => {
```
Comment (picollomartin) [inline app/models/user.js]: This creation of promise is unnecessary, use promise of `User.create` instead

File: app/middlewares/signUpValidations.js [hunk 14]
```diff
@@ -0,0 +1,22 @@
+const { check, validationResult } = require('express-validator/check'),
+  { validateUser } = require('./validations'),
+  error = require('../errors');
+
+exports.signUpValidate = (req, res, next) => {
```
Comment (picollomartin) [inline app/middlewares/signUpValidations.js]: This function should be generic for avoid repeat code in the future

File: app/errors.js [hunk 15]
```diff
@@ -5,3 +5,9 @@ const internalError = (message, internalCode) => ({
 
 exports.DEFAULT_ERROR = 'default_error';
 exports.defaultError = message => internalError(message, exports.DEFAULT_ERROR);
+
+exports.DATABASE_ERROR = 'database_error';
+exports.databaseError = message => internalError(message, exports.DATABASE_ERROR);
+
+exports.SAVING_ERROR = 'saving_error';
+exports.saving_error = message => internalError(message, exports.SAVING_ERROR);
```
Comment (picollomartin) [inline app/errors.js]: Use camelCase for functions and variables

File: app/middlewares/signUpValidations.js [hunk 16]
```diff
@@ -0,0 +1,22 @@
+const { check, validationResult } = require('express-validator/check'),
```
Comment (picollomartin) [inline app/middlewares/signUpValidations.js]: Remove dependency if unused, from `package.json` too with `npm`

File: app/middlewares/signUpValidations.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+        lastName: req.body.lastName,
+        email: req.body.email,
+        password: req.body.password
+      }
+    : {};
+
+  const signErrors = validateUser(user);
+
+  if (!signErrors.valid) {
+    next(error.saving_error(signErrors.messages));
```
Comment (picollomartin) [inline app/middlewares/signUpValidations.js]: Check this when change case of `saving_error`, see next comments

**Review by picollomartin — CHANGES_REQUESTED** [review]
Please request me a review and assign me, for a quick review 😉

File: app/controllers/users.js (hunk 12 above)
Comment (jxczz2) [reply app/controllers/users.js]: the additional properties are ignored and are not a problem, but in theory would be a bad request should you control that?

Comment (picollomartin) [reply app/controllers/users.js]: Not as bad request, but sanitize object before insert to database

File: app/middlewares/signUpValidations.js (hunk 14 above)
Comment (jxczz2) [reply app/middlewares/signUpValidations.js]: I do not understand, is the function incorrect or the name?

Comment (picollomartin) [reply app/middlewares/signUpValidations.js]: The function not the name, should be more generic

File: app/controllers/users.js [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+
+exports.userCreate = (req, res, next) => {
+  const user = {
+    name: req.body.name,
+    lastName: req.body.lastName,
+    email: req.body.email,
+    password: req.body.password
+  };
+
+  User.createUser(user)
```
Comment (picollomartin) [inline app/controllers/users.js]: Return promise here

File: app/middlewares/validations.js [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+const validatePassword = user => {
+  if (!/^(?=\w*\d)(?=\w*[A-Z])(?=\w*[a-z])\S{8,}$/.test(user.password))
+    return { valid: false, message: 'invalid password' };
+  return { valid: true };
+};
+
+const validateMissingValues = user => {
+  for (const member in user) {
+    if (user[member] === '') return { message: 'missing value' };
```
Comment (picollomartin) [inline app/middlewares/validations.js]: What happen if is undefined here ?

File: app/middlewares/signUpValidations.js (hunk 14 above)
Comment (picollomartin) [reply app/middlewares/signUpValidations.js]: Change name too

**Review by picollomartin — CHANGES_REQUESTED** [review]

**Review by picollomartin — APPROVED** [review]
Check some name of variables, things like `index` for object keys

**Pull request merged**